src/item.h

```diff
@@ -206,19 +210,265 @@ class ItemAttributes
 			return static_cast<ItemDecayState_t>(getIntAttr(ITEM_ATTRIBUTE_DECAYSTATE));
 		}
 
+		struct CustomAttributeKey
+		{
+			typedef boost::variant<boost::blank, std::string, int64_t> VariantKey;
+			VariantKey value;
+
+
+			CustomAttributeKey() : value(boost::blank()) {}
+
+			template<typename T>
+			explicit CustomAttributeKey(const T& v) : value(v) {}
+
+			template<typename T>
+			void set(const T& v) {
+				value = v;
+			}
+
+			template<typename T>
+			const T& get();
+
+			struct SerializeVisitor : public boost::static_visitor<> {
+				PropWriteStream& propWriteStream;
+
+				SerializeVisitor(PropWriteStream& propWriteStream) : boost::static_visitor<>(), propWriteStream(propWriteStream) {}
+
+				void operator()(const boost::blank& v) const {
+				}
+
+				void operator()(const std::string& v) const {
+					propWriteStream.writeString(v);
+				}
+
+				void operator()(const int64_t& v) const {
+					propWriteStream.write<int64_t>(v);
+				}
+			};
+
+			void serialize(PropWriteStream& propWriteStream) const {
+				propWriteStream.write<uint8_t>(static_cast<uint8_t>(value.which()));
+				boost::apply_visitor(SerializeVisitor(propWriteStream), value);
+			}
+
+			bool unserialize(PropStream& propStream) {
+				// This is hard coded so it's not general, depends on the position of the variants.
+				uint8_t pos;
+				if (!propStream.read<uint8_t>(pos)) {
+					return false;
+				}
+
+				switch (pos) {
+					case 1:  { // std::string
+						std::string tmp;
+						if (!propStream.readString(tmp)) {
```

Comment:
Isn't it possible to read directly into value instead of a temporary? Same thing for the int64_t.